Add basic Types and remove index prop

coysellers opened the pull request.

coysellers commented:
@alexandervalencia Here's a small PR with the beginnings of some TypeScript. This is a small app and I'm just getting my feet wet again with react so feel free to comment on anything, performance, syntax, setup, etc. Thanks for offering to take a look!

alexandervalencia commented on src/components/CreateTask.tsx:
```diff
@@ -1,6 +1,10 @@
 import React, { useState } from 'react'
 
-export default function CreateTask({ addTask }) {
+type Props = {
```
🚫 For simple component props, we should use an interface since `props` is just an object containing our data for the component.

> Interfaces are basically a way to describe data shapes, for example, an object.
>
> Type is a definition of a type of data, for example, a union, primitive, intersection, tuple, or any other type.

— From LogRocket's solid [post on types vs interfaces](https://blog.logrocket.com/types-vs-interfaces-in-typescript/)

🎸 One thing that may change from person-to-person or company-to-company, is the naming of interfaces/types. Generally from what I've seen, the props name will match the name of the component, so in this case, `CreateTaskProps`.

```suggestion
interface CreateTaskProps {
```

alexandervalencia commented on src/components/CreateTask.tsx:
```diff
@@ -1,6 +1,10 @@
 
-export default function CreateTask({ addTask }) {
+type Props = {
+  addTask: Function
```
From the eslint/recommended settings:
> Don't use `Function` as a type. The `Function` type accepts any function-like value.
> It provides no type safety when calling the function, which can be a common source of bugs.
> It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
>
> If you are expecting the function to accept certain arguments, you should explicitly define the function shape.

Here's how to be more explicit with the shape of the function:
```suggestion
  addTask: (title: string) => void;
```

alexandervalencia commented on src/components/CreateTask.tsx:
```diff
@@ -1,6 +1,10 @@
+  addTask: Function
+}
+
+export const CreateTask = ({ addTask }: Props) => {
```
Don't forget to update the Props interface name here. And depending on how strict you want your typechecking to be, you'll want to add explicit return types. In the case of a React element, `JSX.Element` does the job.

For more (and other good suggestions), see the [React/Typescript cheatsheet](https://github.com/typescript-cheatsheets/react#function-components)


```suggestion
export const CreateTask = ({ addTask }: CreateTaskProps): JSX.Element => {
```

alexandervalencia commented on src/components/Task.tsx:
```diff
@@ -1,8 +1,20 @@
+  task: Task,
+  completeTask: Function,
+  removeTask: Function
+}
```
Same treatment as above:
```suggestion
interface TaskProps {
  task: Task;
  completeTask: (task: Task) => void;
  removeTask: (key: string) => void;
}
```

alexandervalencia commented on src/components/Task.tsx:
```diff
@@ -1,8 +1,20 @@
+interface Task {
+  title: string,
+  completed: boolean,
+  key: string
```
🎸  nitpick

```suggestion
  title: string;
  completed: boolean;
  key: string;

alexandervalencia commented on src/components/Task.tsx:
```diff
@@ -18,8 +30,8 @@ export default function Task({ task, completeTask, removeTask }) {
-            ? "todo-listText text-[#d1d5db]"
+          className={task.completed ?
+            "todo-listText text-[#d1d5db]"
             : "todo-listText"
```
Since `todo-listText` should be included either way, let's make that clear and only swap out the color change based on completed status:
```suggestion
          className={`todo-listText ${task.completed ? "text-[#d1d5db]" : ""}`}
```


This is also where the fantastic [classnames package](https://github.com/JedWatson/classnames) would be really useful—but probably overkill for this single usage.

e.g.
```js
import classNames from "classnames"

className={classNames("todo-listText",  {"text-[#d1d5db]": task.completed}}
```

alexandervalencia commented on src/components/Task.tsx:
```diff
@@ -1,8 +1,20 @@
+  removeTask: Function
+}
+
+export const Task = ({ task, completeTask, removeTask }: Props) => {
```
```suggestion
export const Task = ({ task, completeTask, removeTask }: TaskProps): JSX.Element => {
```

alexandervalencia commented on src/components/Task.tsx:
```diff
@@ -1,8 +1,20 @@
+
+export const Task = ({ task, completeTask, removeTask }: Props) => {
+  const checkMark = task.completed ?
+    "m8 13 2.165 2.165a1 1 0 0 0 1.521-.126L16 9"
```
This was correct before
```suggestion
  const checkMark = task.completed
    ? "m8 13 2.165 2.165a1 1 0 0 0 1.521-.126L16 9"
```

alexandervalencia replied:
I'm not sure if you're maybe using a plugin to do linting/formatting? But I would recommend adding those dev dependencies to the project and letting the linters do the work.

Here's a basic start:

```
yarn add --dev eslint prettier eslint-plugin-prettier eslint-config-prettier
npx eslint --init
```

```jsonc
// package.json
"scripts": {
  "lint": "eslint src --ext .js,.tsx --fix",
  "format": "prettier --ignore-path .gitignore --write \"**/*.+(js|ts|tsx|json)\"",
}
```

[Another decent article by LogRocket](https://blog.logrocket.com/linting-typescript-using-eslint-and-prettier/) (though you shouldn't need much more than what's here).

And don't forget to set up the `.eslintrc.js` to account for prettier rules:
```js
// .eslintrc.js
{
  // ...
  extends: [
    "eslint:recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:prettier/recommended",
  ],
  plugins: ["react", "@typescript-eslint", "prettier"],
```

alexandervalencia reviewed (approved):
Hey Coy! Thanks for trusting me to provide feedback! I know this particular PR was about adding basic types for the React components but I had some other thoughts I wanted to give...

1) I think this might be more of a personal preference but for the Task component, we are handing over a `task` prop which contains the `key`, `completed`, and `title` properties. I feel like I would prefer that instead of requiring a `task`, the component just required the properties themselves. But again, if you disagreed I wouldn't think you should change it.

2) We have two handlers (`completeTask` & `removeTask`) that are similar in actions (modifying the state of the task) but using different arguments (`task` & `key` (the `task.key`) respectively). I think if we could bring them more in line it would make things clearer and cleaner.

A "React standard" to consider: `completeTask` is currently operating in a "Vue" like way where we are directly mutating state. But for React, we shouldn't be modifying state outside of `setNameOfWhateverState()` type hooks. So I would advise updating the function to: a) use only the key param and b) ensure we're making a true clone and not mutating existing data.

Right now, because we're spreading the contents of `tasks` to a new array, we aren't making copies of the objects, just the references _to_ them. (thanks JavaScript 🙄 )

Here's one approach you could take:
```js
const completeTask = ({ key }) => {
  const updatedTasks = tasks.map((task) => ({ ...task }));
  const completedTask = updatedTasks.find((task) => task.key === key);

  completedTask.completed = !completedTask.completed;

  setTasks([...updatedTasks]);
};
```

Let me know if you have any questions! The app itself looks really nice and I'm stoked you're working on growing with this stuff!

coysellers commented:
@alexandervalencia Thanks for all the feed back!

First of all, I didn't know I was mutating the sources here:

```js
const updatedTasks = [...tasks];

key.completed = !key.completed
```
That's good to know.

Although I'm having a little trouble understanding what's going on with filter here:
```js
const completeTask = ({ key }) => {
    const updatedTasks = tasks.map((task) => ({ ...task }));
    // console.log(updatedTasks);
    // 0: {title: 'asdfl', completed: true, key: '79ebdf3e-bd58-4e6c-a0d3-9771edffcc5e'}
    // 1: {title: 'adslkfjasdf', completed: false, key: 'b8073295-984b-4b92-93b4-edb110d72053'}

    const completedTask = updatedTasks.find((task) => task.key === key);
    // console.log(completedTask);
    // {title: 'asdfl', completed: false, key: '79ebdf3e-bd58-4e6c-a0d3-9771edffcc5e'}
  
    completedTask.completed = !completedTask.completed;
  
    setTasks([...updatedTasks]);
};
```

I understand what this is accomplishing, but when I log `updatedTasks` (by clicking the check mark box) I get the new state of the object `completed: true`. But when I log `completedTask` it returns the opposite `completed: false` and I'm not sure why that is.

We are mapping through the original tasks and spreading each task into new Objects, filtering through and finding the keys that match, changing the boolean of the one that matches, and then spreading the new objects into and array and setting the state, right?

Which part is creating a copy rather than using a reference, is it the spreading of tasks `{ ...task }` or setting the state in the array `[...updatedTasks]`?

coysellers commented:
@alexandervalencia Also for the Task component I'm having a little trouble understanding that one too.

Are you saying somehow pass in the task object directly into the component, roughly like this?:
```js
<Task
    task={{
        title,
        completed: false,
        key: uuidv4()
    }}
    key={task.key}
    completeTask={completeTask}
/>
```

Or did you mean passing it in somehow in the component file?
```js
export const Task = ({ task, completeTask, removeTask }) => {...
```

alexandervalencia commented:
Hey @coysellers for the Task component here is what I mean...

Lets say we define a singular task data as follows:
```typescript
interface Task {
  completed: boolean;
  key: string;
  title: string;
}
```

A React component should essentially represent a piece of the UI. So for my `Task` component, I would expect it to adhere to the same definition I've given the `Task` data. It should take: `key`, `completed`, and `title`.

```tsx
const [tasks, setTasks] = useState<Task[]>([
  {
    completed: false,
    key: "task-1",
    title: "make a todo list",
  }
]);

// further down while rendering...

{tasks.map(({ completed, key, title }) => (
  <Task
    completed={completed}
    key={key}
    title={title}
  />
))}
```

As opposed to the idea that a Task component receives a "task" object.

```tsx
{tasks.map((task) => <Task key={task.key} task={task} />)}
```

Like I mentioned before, I think _this_ particular idea is more a matter of personal opinion and I can imagine reasons why you would do it the second way over the first and the standards will most likely differ from team to team, depending on who you're working with. 😁 


One last note... if we went with the first way we would also need to adjust the `<Task />` to accept the correct props...

```tsx
// Task.tsx

interface TaskProps {
  title: string;
  completed: boolean;
  key: string;
  completeTask: (task: Task) => void;
  removeTask: (key: string) => void;
}

export const Task = ({
  completed,
  completeTask,
  key,
  removeTask,
  title,
}: TaskProps): JSX.Element => {
  // ... omitted for example
}
```

alexandervalencia commented:
> Although I'm having a little trouble understanding what's going on with filter here:
> 
> ```js
> const completeTask = ({ key }) => {
>     const updatedTasks = tasks.map((task) => ({ ...task }));
>     // console.log(updatedTasks);
>     // 0: {title: 'asdfl', completed: true, key: '79ebdf3e-bd58-4e6c-a0d3-9771edffcc5e'}
>     // 1: {title: 'adslkfjasdf', completed: false, key: 'b8073295-984b-4b92-93b4-edb110d72053'}
> 
>     const completedTask = updatedTasks.find((task) => task.key === key);
>     // console.log(completedTask);
>     // {title: 'asdfl', completed: false, key: '79ebdf3e-bd58-4e6c-a0d3-9771edffcc5e'}
>   
>     completedTask.completed = !completedTask.completed;
>   
>     setTasks([...updatedTasks]);
> };
> ```
> 
> I understand what this is accomplishing, but when I log `updatedTasks` (by clicking the check mark box) I get the new state of the object `completed: true`. But when I log `completedTask` it returns the opposite `completed: false` and I'm not sure why that is.

Okay, I noticed that weirdness too, `console.log(updatedTasks)` is logging the _updated_ state when it seems like it should be the _current_ state of the tasks. I think it's a weird timing issue with the way the functions/rendering is being called? But when I did `console.log(...updatedTasks)` the tasks logged out the correct state. 🤷‍♂️ 

> We are mapping through the original tasks and spreading each task into new Objects, filtering through and finding the keys that match, changing the boolean of the one that matches, and then spreading the new objects into and array and setting the state, right?

Right!

> Which part is creating a copy rather than using a reference, is it the spreading of tasks `{ ...task }` or setting the state in the array `[...updatedTasks]`?

Sorry for this being extra confusing!
First let me clear something up... spreading out the tasks at the end was completely unnecessary: `setTasks([...updatedTasks]);` should just be `setTasks(updatedTasks);`

Yes, when we map over the existing tasks and spread out the contents into a new object, that's when we are creating copies instead of just referencing the existing objects in memory. Keep in mind that if tasks held another object, e.g. 
```js
{ key: '1', completed: false, title: 'walk the dog', options: { repeat: true, favorite: true }
```
We would need to ALSO spread out _that_ object, otherwise, we are copying the *reference* to the object by spreading the object contents: `{ ...task }`. So instead we would need to do this: `{ ...task, options: { ...task.options } }`
And that's why folks advise to keep state as shallow as possible 😁 


About `completedTask.completed = !completedTask.completed`... Yes, we're still mutating data, but now we're mutating our the object inside our temporary `updatedTasks` instead of the current `tasks`.

One other thought I had is to use reduce to do the copying, filtering, and updating, all at once!

```ts
const completeTask = (key: string) => {
  const updatedTasks = tasks.reduce<Task[]>((newTasks, task) => {
    const newTask = {
      ...task,
      completed: task.key === key ? !task.completed : task.completed,
    };
    return [...newTasks, newTask];
  }, []);

  setTasks(updatedTasks);
};
```

alexandervalencia commented:
One final thought... I know you're just barely getting started learning/integrating TS with your project, but don't forget to eventually add a `tsconfig.json` so you'll get all that proper type-checking and error feedback. You'll need to update your `package.json` as well for your build step and grab a few more packages.

I highly suggest checking out this repo as a starter template: https://github.com/uchihamalolan/vite-react-ts

(You don't need to start over, you can just copy the files you need when you're ready)

coysellers commented:
@alexandervalencia So I made a majority of the changes you suggested but I'm still struggling a bit with the `"A React component should essentially represent a piece of the UI"` concept. Are there any good resources around this?

I changed that out and mostly got it working, though I am still running into some issues with the `key` prop. When I used this `key={key}` in the Task component it breaks and I get a warning `Warning: Task: `key` is not a prop. Trying to access it will result in `undefined` being returned.`.

So instead I renamed the prop like this, `itemId={key}`. That fixed the undefined issue but now I'm getting a `Warning: Each child in a list should have a unique "key" prop.` which is a bit confusing but I think I am passing in a valid key to each list item. Maybe because it's called something else (`itemId`) in the Task object being passed in so it can't reconcile the two? I'm not really sure. I even put a `data-item-key={itemId}` attribute on the `li` to make sure the `itemId` was getting properly assigned. I pushed up the partially working changes if you have time to take a look.

Thanks again! This is good stuff!

alexandervalencia commented:
@coysellers Sorry it took forever to get back to you!

> I'm still struggling a bit with the "A React component should essentially represent a piece of the UI" concept. Are there any good resources around this?

So I asked around folks at work and got mixed responses on this. I started searching around and I couldn't really find a good resource, so I'm going to chalk this one up to me learning something that was a personal preference but somehow posed as, "it should only be done this way" in my mind.

Regarding the `key` issue you were running into, keys must exist on the component being called in the `map()`. [See this react doc](https://reactjs.org/docs/lists-and-keys.html#extracting-components-with-keys) for exact details on the same problem you were running into.
